src/consensus/params.cpp

```diff
@@ -36,32 +39,153 @@ namespace Consensus {
+                - (BLOSSOM_POW_TARGET_SPACING_RATIO * (blossomActivationHeight - SubsidySlowStartShift()))
+                + blossomActivationHeight;
         } else {
-            return nPreBlossomSubsidyHalvingInterval + SubsidySlowStartShift() - 1;
+            return (nPreBlossomSubsidyHalvingInterval * halvingIndex) + SubsidySlowStartShift();
         }
     }
 
+    int Params::GetLastFoundersRewardBlockHeight(int nHeight) const {
+        return HalvingHeight(nHeight, 1) - 1;
+    }
+
+    int Params::FundingPeriodIndex(int fundingStreamStartHeight, int nHeight) const {
+        int firstHalvingHeight = HalvingHeight(fundingStreamStartHeight, 1);
+
+        // If the start height of the funding period is not aligned to a multiple of the
+        // funding period length, the first funding period will be shorter than the
+        // funding period length.
+        auto startPeriodOffset = (fundingStreamStartHeight - firstHalvingHeight) % nFundingPeriodLength;
+        if (startPeriodOffset < 0) startPeriodOffset += nFundingPeriodLength; // C++ '%' is remainder, not modulus!
+
+        return (nHeight - fundingStreamStartHeight + startPeriodOffset) / nFundingPeriodLength;
+    }
+
+    boost::variant<FundingStream, FundingStreamError> FundingStream::ValidateFundingStream(
+        const Consensus::Params& params,
+        const int startHeight,
+        const int endHeight,
+        const std::vector<FundingStreamAddress>& addresses
+    ) {
+        if (!params.NetworkUpgradeActive(startHeight, Consensus::UPGRADE_CANOPY)) {
+            return FundingStreamError::CANOPY_NOT_ACTIVE;
+        }
+
+        if (endHeight < startHeight) {
+            return FundingStreamError::ILLEGAL_RANGE;
+        }
+
+        if (params.FundingPeriodIndex(startHeight, endHeight - 1) >= addresses.size()) {
+            return FundingStreamError::INSUFFICIENT_ADDRESSES;
+        }
+
+        return FundingStream(startHeight, endHeight, addresses);
+    };
+
+    class GetFundingStreamOrThrow: public boost::static_visitor<FundingStream> {
+    public:
+        FundingStream operator()(const FundingStream& fs) const {
+            return fs;
+        }
+
+        FundingStream operator()(const FundingStreamError& e) const {
+            switch (e) {
+                case FundingStreamError::CANOPY_NOT_ACTIVE:
+                    throw std::runtime_error("Canopy network upgrade not active at funding stream start height.");
+                case FundingStreamError::ILLEGAL_RANGE:
+                    throw std::runtime_error("Illegal start/end height combination for funding stream.");
+                case FundingStreamError::INSUFFICIENT_ADDRESSES:
+                    throw std::runtime_error("Insufficient payment addresses to fully exhaust funding stream.");
+                default:
+                    throw std::runtime_error("Unrecognized error validating funding stream.");
+            };
```

Comment:
Is this semi-colon needed here? Same after the closing brace of several of the class methods here?

Reply:
Indeed those semicolons are not needed.